packages/jubmoji-api/src/proofs.ts

```diff
@@ -323,3 +328,130 @@ export class PublicMessageSignature
     });
   }
 }
+
+export class TeamLeaderboard
+  implements ProofClass<TeamLeaderboardProofArgs, TeamLeaderboardProof>
+{
+  teamPubKeys: string[];
+  collectionPubKeys: string[];
+  sigNullifierRandomness: string;
+  pathToCircuits?: string;
+
+  constructor({
+    teamPubKeys,
+    collectionPubKeys,
+    sigNullifierRandomness,
+    pathToCircuits,
+  }: TeamLeaderboardClassArgs) {
+    this.teamPubKeys = teamPubKeys;
+    this.collectionPubKeys = collectionPubKeys;
+    this.sigNullifierRandomness = sigNullifierRandomness;
+    this.pathToCircuits = pathToCircuits;
+  }
+
+  async prove({
+    teamJubmoji,
+    collectionJubmojis,
+  }: TeamLeaderboardProofArgs): Promise<TeamLeaderboardProof> {
+    const pubKeyNullifierRandomness = hexToBigInt(
+      getRandomNullifierRandomness()
+    );
+
+    const { sig, msgHash, pubKey, R, T, U } =
+      getMembershipProofArgsFromJubmoji(teamJubmoji);
+    const teamPubKey = getCardPubKeyFromIndex(teamJubmoji.pubKeyIndex);
+    // We need to reveal which team we are on for the leaderboard
+    const teamMerkleProof = await getMerkleProofFromCache([teamPubKey], 0);
+
+    const teamMembershipProof = await proveMembership({
+      sig,
+      msgHash,
+      publicInputs: {
+        R,
+        T,
+        U,
+      },
+      merkleProof: teamMerkleProof,
+      sigNullifierRandomness: hexToBigInt(this.sigNullifierRandomness),
+      pubKeyNullifierRandomness,
+      pathToCircuits: this.pathToCircuits,
+    });
+
+    const sigs = [];
+    const msgHashes = [];
+    const publicInputs = [];
+    const indices = [];
+    for (const jubmoji of collectionJubmojis) {
```

Comment:
Are we tracking an issue where we can avoid recomputing proofs for ones we've already done? / Building some sort of localStorage stash

Reply:
https://github.com/jubmoji/jubmoji.quest/issues/67